File: ComplainstApp/app/src/main/java/com/example/complainstapp/LoginActivity.java
package com.example.complainstapp;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

import com.androidnetworking.AndroidNetworking;
import com.androidnetworking.common.Priority;
import com.androidnetworking.error.ANError;
import com.androidnetworking.interfaces.JSONArrayRequestListener;
import com.androidnetworking.interfaces.JSONObjectRequestListener;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.util.Log;
import android.view.View;

import com.google.android.material.textfield.TextInputLayout;

import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class LoginActivity extends AppCompatActivity {

    private String accessToken;
    private TextInputLayout idLayout;
    private TextInputLayout passwordLayout;
    private Button loginButton;
    private TextView signUpButton;
    private ImageButton backButton;
    GoogleSignInClient mGoogleSignInClient;
    ImageButton signInButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        AndroidNetworking.initialize(getApplicationContext());

        idLayout = findViewById(R.id.outlinedIdField);
        passwordLayout = findViewById(R.id.outlinedPassField);
        loginButton = findViewById(R.id.button);
        signUpButton = findViewById(R.id.textView4);
        backButton = findViewById(R.id.imageButton);

        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken("689285763404-9ih3lrpb9154mhob4rs8oqbpruvng95s.apps.googleusercontent.com")
                .requestEmail()
                .build();

        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);

        signInButton = findViewById(R.id.signInButton);

        signInButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent signInIntent = mGoogleSignInClient.getSignInIntent();
                startActivityForResult.launch(signInIntent);
            }
        });



        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(!validateEmail() | !validatePassword()){
                    return;
                }

                AndroidNetworking.post("http://192.168.0.109:5000/login")
                        .addBodyParameter("nsuid", idLayout.getEditText().getText().toString())
                        .addBodyParameter("password", passwordLayout.getEditText().getText().toString())
                        .setTag("test")
                        .setPriority(Priority.HIGH)
                        .build()
                        .getAsJSONObject(new JSONObjectRequestListener() {
                            @Override
                            public void onResponse(JSONObject response) {
                                try {
                                    accessToken = response.getString("accessToken");
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                }
                                Intent intent = new Intent(LoginActivity.this, HomepageActivity.class);
                                intent.putExtra("token",accessToken);
                                startActivity(intent);
                            }
                            @Override
                            public void onError(ANError error) {
                                if(error.getErrorCode()==401) {
                                    Toast.makeText(LoginActivity.this, "Password is incorrect!", Toast.LENGTH_SHORT).show();
                                }
                                else if(error.getErrorCode()==512){
                                    Toast.makeText(LoginActivity.this, "User is not verified!", Toast.LENGTH_SHORT).show();
                                }
                                else{
                                    Log.e("Error", String.valueOf(error.getErrorCode()));
                                    Toast.makeText(LoginActivity.this, "An error occurred.Try again!", Toast.LENGTH_SHORT).show();
                                }
                            }
                        });

            }
        });

        signUpButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Uri uri = Uri.parse("http://192.168.0.109:3000/signup");
                Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                startActivity(intent);
            }
        });

        backButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                finish();
            }
        });

    }

    ActivityResultLauncher<Intent> startActivityForResult = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult result) {
                    if(result.getResultCode() == Activity.RESULT_OK){
                        Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(result.getData());
                        handleSignInResult(task);
                    }
                }
            }
    );

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount acct = completedTask.getResult(ApiException.class);
            if (acct != null) {
                String personName = acct.getDisplayName();
                String personGivenName = acct.getGivenName();
                String personFamilyName = acct.getFamilyName();
                String personEmail = acct.getEmail();
                String personId = acct.getId();
                Uri personPhoto = acct.getPhotoUrl();

                Log.e("Token Id",acct.getIdToken());
                Log.e("Acct Id",acct.getDisplayName().substring((acct.getDisplayName().lastIndexOf(" ") + 1)));

                AndroidNetworking.post("http://192.168.0.109:5000/Gsignup")
                        .addBodyParameter("token",acct.getIdToken())
                        .addBodyParameter("nsuid",acct.getDisplayName().substring((acct.getDisplayName().lastIndexOf(" ") + 1)))
                        .setTag("test1")
                        .setPriority(Priority.HIGH)
                        .build()
                        .getAsJSONObject(new JSONObjectRequestListener(){
                            @Override
                            public void onResponse(JSONObject response) {
                                try {
                                    accessToken = response.getString("accessToken");
                                } catch (JSONException e) {
                                    e.printStackTrace();
                                }
                                Intent myIntent = new Intent(LoginActivity.this, HomepageActivity.class);
                                myIntent.putExtra("token",accessToken);
                                startActivity(myIntent);
                                finish();
                            }
                            @Override
                            public void onError(ANError error) {
                                Log.e("google error",error.toString());
                            }
                        });

            }else {
                // Signed out, show unauthenticated UI.
                Toast.makeText(this,"Account does not exist. Please sign up!",Toast.LENGTH_LONG).show();
            }

        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.d("GOOGLE ERROR", e.getMessage());
        }
    }

    private boolean validateEmail(){
        String emailInput = idLayout.getEditText().getText().toString().trim();

        if(emailInput.isEmpty()){
            idLayout.setError("Field can't be empty!");
            return false;
        }
        else if(!(emailInput.length()==10)){
            idLayout.setError("NSU id must be 10 characters long!");
            return false;
        }
        else{
            idLayout.setError(null);
            return true;
        }
    }

    private boolean validatePassword(){
        String passInput = passwordLayout.getEditText().getText().toString().trim();

        if(passInput.isEmpty()){
            passwordLayout.setError("Field can't be empty!");
            return false;
        }
        else{
            passwordLayout.setError(null);
            return true;
        }
    }

}

File: CSE327/app/src/main/java/com/example/cse327/LoginActivity.kt
package com.example.cse327

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task


const val RC_SIGN_IN = 123

class LoginActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        var checkError = false;

        val signUpButton = findViewById<TextView>(R.id.textView4);
        val imageButton2 = findViewById<ImageButton>(R.id.imageButton2);
        val emailField = findViewById<EditText>(R.id.emailField)
        val passField = findViewById<EditText>(R.id.passField)
        val errorUser = findViewById<TextView>(R.id.errorEmail)
        val errorPass = findViewById<TextView>(R.id.errorPass)
        val backButton = findViewById<ImageButton>(R.id.imageButton)
        val loginButton = findViewById<Button>(R.id.button)


        // set on-click listener
        signUpButton.setOnClickListener{
            val url = "http://www.facebook.com"
            val i = Intent(Intent.ACTION_VIEW)
            i.data = Uri.parse(url)
            startActivity(i)
        }

        // set on-click listener
        backButton.setOnClickListener {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
        }

        // set on-click listener
        loginButton.setOnClickListener {
            val url : String = emailField.text.toString();
            val check : Boolean = "@northsouth.edu" in url
            if(check==false){
                errorUser.setText("Please use an NSU email.")
                checkError = true;
            }
            if(url.trim().isEmpty()){
                errorUser.setText("Please enter an email.")
                checkError = true;
            }
            if(passField.text.toString().length < 6) {
                errorPass.setText("Password is too small.")
                checkError = true;
            }
            if(passField.text.toString().trim().isEmpty()) {
                errorPass.setText("Please enter a password.")
                checkError = true;
            }
            if(checkError==false) {
                val intent = Intent(this, HomepageActivity::class.java)
                startActivity(intent)
            }
        }

        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .build()
        val mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        imageButton2.visibility = View.VISIBLE;
        imageButton2.setOnClickListener{
            val signInIntent = mGoogleSignInClient.signInIntent
            startActivityForResult(signInIntent, RC_SIGN_IN)
        }

        val acct = GoogleSignIn.getLastSignedInAccount(this)
        if (acct != null) {
            val personName = acct.displayName
            val personGivenName = acct.givenName
            val personFamilyName = acct.familyName
            val personEmail = acct.email
            val personId = acct.id
            val personPhoto = acct.photoUrl
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        // Result returned from launching the Intent from GoogleSignInClient.getSignInIntent(...);
        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.
            val task = GoogleSignIn.getSignedInAccountFromIntent(data)
            handleSignInResult(task)
        }
    }

    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val account = completedTask.getResult(ApiException::class.java)
            // Signed in successfully, show authenticated UI.
            val intent = Intent(this, HomepageActivity::class.java)
            startActivity(intent)
        } catch (e: ApiException) {
            Toast.makeText(getApplicationContext(),"Google Sign-In unsuccessful!", Toast.LENGTH_SHORT).show();
        }
    }

}